draft timeline viz doc (#18788)
DataDog/documentation: content/en/profiler/profile_visualizations.md

      text: 'Introducing always-on production profiling in Datadog'
---

> ## Search profiles

Author: It is not an overview anymore, given the page is now focused on visualizations. This part is solely dedicated to explain how to search profiles (before visualizing them)

For each runtime, there is also a broader set of metrics available, which you can see [listed by timeseries][3].

## Profile types

In the **Profiles** tab, you can see all profile types available for a given language. Depending on the language, the information collected about your profile differs. See [Profile types][4] for a list of profile types available for each language.

## Visualizations

### Flame Graph

Continuous Profiler default visualization is the flame graph.

{{< img src="profiler/profiling_viz-flamegraph.png" alt="A flame graph" >}}

It shows are how much CPU each method used (since this is a CPU profile) and how each method was called.

For example, starting from the first row, `Thread.run()` called `ThreadPoolExecutor$Worker.run()` which called `ThreadPoolExecutor.runWorker(ThreadPoolExecutor$Worker)` and so on.

The width of a frame represents how much of the total CPU it consumed. On the right, you can see a "CPU time by Method" top list that only accounts for self time, which is the time a method spent on CPU without calling another method.

> ### Single Profile

Reviewer: ```suggestion
### Single profile
```

By default, profiles are uploaded once a minute. Depending on the language, these processes are profiled between 15s and 60s.

> To view a specific profile, select `Visualize as Profile list` and click on a line to view a specific profile:

Reviewer: ```suggestion
To view a specific profile, set the **Visualize as** option to **Profile List** and click an item in the list:
```

**Note**: In the upper right corner of each profile, there are options to:

- Compare this profile with others
- View repository commit
- View traces for the same process and timeframe
- Download the profile
- Open the profile in full page

### Timeline view

The timeline view is the equivalent of the flame graph, with a distribution over time.

{{< img src="profiler/profiling_viz-timeline.png" alt="A timeline" >}}

It shows time-based patterns and work distribution over the period of a single profile.

Compared to the flame graph, the timeline view can help you:

- Isolate spiky methods
- Sort out complex interactions between threads
- Surface runtime activity that impacted the process

Depending on the runtime and language, the timeline lanes vary:

{{< programming-lang-wrapper langs="java,go,dotnet" >}}
{{< programming-lang lang="java" >}}

> Each lane is a **thread**. Threads from a common pool are grouped together. You can expand the pool to see each thread details.

Reviewer: ```suggestion
Each lane represents a **thread**. Threads from a common pool are grouped together. You can expand the pool to view details for each thread.
```

> Lanes are top are runtime activities that may impact performance
> {{< /programming-lang >}}
> {{< programming-lang lang="go" >}}
> This feature needs to be explicitly enabled for Go, see [prerequisites][1].

Reviewer: ```suggestion
Lanes on top are runtime activities that may impact performance.
{{< /programming-lang >}}
{{< programming-lang lang="go" >}}
See [prerequisites][1] to learn how to enable this feature for Go.
```

> Each lane is a **goroutine**. Goroutines created by the same `go` statement are grouped together. You can expand the group to see each goroutine's details.

Reviewer: ```suggestion
Each lane represents a **goroutine**. Goroutines created by the same `go` statement are grouped together. You can expand the group to view details for each goroutine.
```

Lanes on top are runtime activities that may impact performance.

[1]: /profiler/connect_traces_and_profiles/#prerequisites
{{< /programming-lang >}}
{{< programming-lang lang="dotnet" >}}

> Each lane is a **thread**. Threads from a common pool are grouped together. You can expand the pool to see each thread details.

Reviewer: ```suggestion
Each lane represents a **thread**. Threads from a common pool are grouped together. You can expand the pool to view details for each thread.
```

> Lanes are top are runtime activities that may impact performance.

Reviewer: ```suggestion
Lanes on top are runtime activities that may impact performance.
```
